### PR TITLE
Make the Sandbox truly pluggable

### PR DESCRIPTION
**Pull request by tlvenn**
As discussed with @josevalim in #1959, this small PR removes the hardcoded references on `Ecto.Adapters.SQL.Sandbox` so that one can plug his own version of the sandbox without overriding the existing `Ecto.Adapters.SQL.Sandbox`.

### CODE REVIEW

File: lib/ecto/adapters/sql.ex [hunk 1]
```diff
@@ -317,10 +317,9 @@ defmodule Ecto.Adapters.SQL do
     |> Keyword.put_new(:pool_timeout, @pool_timeout)
   end
 
-  defp normalize_pool(Ecto.Adapters.SQL.Sandbox),
-    do: DBConnection.Ownership
-  defp normalize_pool(pool),
-    do: pool
+  defp normalize_pool(pool) do
+    if function_exported?(pool, :unboxed_run, 2), do: DBConnection.Ownership, else: pool
```
Comment (josevalim) [inline lib/ecto/adapters/sql.ex]: I don't think we should do this change. We only do it for backwards compatibility, your sandbox should require DBConnection.Ownership to be explicitly passed instead.

Comment (tlvenn) [inline lib/ecto/adapters/sql.ex]: Ha I didn't know about that, I thought it was some implicit contract, reverted to the original code.

File: lib/ecto/adapters/sql.ex [hunk 2]
```diff
@@ -348,11 +348,10 @@ defmodule Ecto.Adapters.SQL do
 
     # Check if the pool options should overridden
     {pool_name, pool_opts} =
-      case Keyword.fetch(opts, :pool) do
```
Comment (josevalim) [inline lib/ecto/adapters/sql.ex]: I believe this case is outdated. I have remove it from master.

Comment (tlvenn) [inline lib/ecto/adapters/sql.ex]: Ha thanks fixed.

**Comment by tlvenn** [comment]
I have pushed a branch for my sandbox to test out how it works on its own.
> https://github.com/jumpn/cockroachdb_sandbox/tree/no-override

I had to provide `ensure_all_started` and `child_spec` and I am not entirely convinced I am doing it properly. Would very much appreciate some guidance there.

And beyond that, my sandbox does not work any more:

```
** (UndefinedFunctionError) function Jumpn.Repo.Pool.__pool__/0 is undefined (module Jumpn.Repo.Pool is not available)
    Jumpn.Repo.Pool.__pool__()
    lib/cockroachdb_sandbox.ex:586: CockroachDBSandbox.proxy_pool/1
    lib/cockroachdb_sandbox.ex:518: CockroachDBSandbox.checkout/2
    (db_connection) lib/db_connection.ex:919: DBConnection.checkout/2
    (db_connection) lib/db_connection.ex:741: DBConnection.run/3
    (db_connection) lib/db_connection.ex:1132: DBConnection.run_meter/3
    (db_connection) lib/db_connection.ex:584: DBConnection.prepare_execute/4
    (ecto) lib/ecto/adapters/postgres/connection.ex:86: Ecto.Adapters.Postgres.Connection.execute/4
```

For some reason, the repo arg being passed to `checkout` is `Jumpn.Repo.Pool` all of the sudden instead of the expected `Jumpn.Repo`

**Comment by josevalim** [comment]
@tlvenn your pool should be `pool: DBConnection.Ownership` and set `ownership_pool: ToYourCurrentPool`.

**Comment by tlvenn** [comment]
Ha make sense, i was a puzzled by that and decided to ignore it, I guess I should not have :)

Also i am curious as to how `__pool__` got injected into my repo, I dont understand how this is happening.

**Comment by josevalim** [comment]
@tlvenn your pool should be `pool: DBConnection.Ownership` and your module is just conveniences for working with the pool, nothing more. :)

**Comment by josevalim** [comment]
@tlvenn the `__pool__` configuration is injected into the repository by the adapter. There is an open issue to make the `__pool__` configuration dynamic and get rid of the `__pool__` compile-time injection.

**Comment by tlvenn** [comment]
Regarding the `__pool__`, it is injected when you declare the following?
```
def __adapter__ do
        @adapter
      end
```

As for the sandbox, i am not sure I actually understand what you are suggesting. If I define the pool to use `DBConnection.Ownership` then all the checks we do in the PR will fail to detect we are using a sandbox. Am i missing something ?

**Comment by tlvenn** [comment]
Solved the issue by changing back the `normalize_pool/1`  to:
```
defp normalize_pool(pool) do
   if function_exported?(pool, :unboxed_run, 2), do: DBConnection.Ownership, else: pool
end
```

Now my sandbox works properly.
So it seems this is needed for more than backward compatibility.

**Comment by josevalim** [comment]
@tlvenn you are right. We need to figure out a better way to handle this then. I am going to merge your PR as is and then try to come up with a better approach than checking for unboxed_run on normalize_config.

**Comment by tlvenn** [comment]
Thanks for the review @josevalim and have a great WE.

**Comment by tlvenn** [comment]
The tests are failing with Travis but I cant seem to reproduce it, all the tests are passing on my computer. I have seen that error before but it disappeared after a `mix clean`. I dont really understand how this is happening given the scope of the changes I made.

The error:
```
** (UndefinedFunctionError) function Ecto.Adapters.SQL.Sandbox.ensure_all_started/2 is undefined or private
    (ecto) Ecto.Adapters.SQL.Sandbox.ensure_all_started([pool_timeout: 5000, timeout: 15000, loggers: [Ecto.LogEntry, {Ecto.Integration.Repo, :log, [:on_log]}], adapter: Ecto.Adapters.Postgres, url: "ecto://postgres:postgres@localhost/ecto_test", pool: Ecto.Adapters.SQL.Sandbox, ownership_pool: DBConnection.Sojourn], :temporary)
    (ecto) lib/ecto/adapters/postgres.ex:86: Ecto.Adapters.Postgres.ensure_all_started/2
    integration_test/pg/test_helper.exs:68: (file)
    (elixir) lib/code.ex:363: Code.require_file/2
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
```

**Comment by josevalim** [comment]
It is happening because `function_exported?` does not load the module. That said, the check you added to normalize_config will never pass, even for the sandbox, because the module has not been loaded yet. So we never rewrite the sandbox pool entry to DBConnection.Ownership.

**Comment by tlvenn** [comment]
Hi @josevalim, can you help out on this one, I am not so intimate with how/when module loading works and how to best address this issue. Thanks in advance.

**Comment by tlvenn** [comment]
I guess a `Code.ensure_loaded(pool)` does the trick

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**